packages/.gitignore

```diff
@@ -0,0 +1,2 @@
+## Ignore composer.lock for our library packages
+/**/composer.lock
```

Comment:
Can this still get generated and can the untracked presence of this file produce different results for different people?

Is there a way to disable it being generated in the first place?

Reply:
> Is there a way to disable it being generated in the first place?

Not to my knowledge.

> Can the untracked presence of this file produce different results for different people?

It can, but that's generally the idea for libraries - they should be flexible and reusable in different contexts. As far as I understand they are expected to produce different results, and that should be just fine. If it's not, then it's highly possible that the library is doing something wrong.

Reply:
Valid point about package dynamics!

Do these dependencies still get locked in Jetpack repo in the root lockfile, then?

If they do, great!

If not, wouldn't we have a situation where Jetpack's dependencies are not consistent for everyone in the monorepo/plugin between developers and the CI? Each time people type `composer install`, they'd get different versions. That's not the case for consumers of packaged versions of these monorepo-packages, who have their own lockfiles and need to type `composer update` to bump versions of transitive deps in packages anyway.

Could the lockfile be stripped out when mirroring these packages to their own repositories? Lockfile could even just be left in and when packages are used as a dependency, the lockfile will be ignored:

> _For your library you may commit the composer.lock file if you want to. This can help your team to always test against the same dependency versions. However, this lock file will not have any effect on other projects that depend on it. It only has an effect on the main project._

([via](https://getcomposer.org/doc/02-libraries.md#lock-file))

There might be angle I'm missing tho but just wanted to make sure deps stay locked in main repo as they should. :-)

Sidenote: this just reminded me of this situation in the past where the main repo didn't have package-lock commited https://github.com/Automattic/newspack-blocks/pull/6#issuecomment-482074977

Reply:
> Do these dependencies still get locked in Jetpack repo in the root lockfile, then?

Yes. The idea is we should have `composer.lock` in the root of Jetpack (as it's a project), but not in the packages (which are libraries).

> but just wanted to make sure deps stay locked in main repo as they should

Yeah, that's understandable, and thanks for checking.  As you can see in this PR, we're committing the root `composer.lock` and we're only ignoring the ones in `packages/*`.

Let me know if you have any other concerns, though!

Reply:
So installed dependency versions from `/packages/*/composer.json` files _do_ get hoisted to the root lockfile? Just want to make sure I understood you correctly.

Reply:
I see where that gets confusing. 

It really depends on how we decide to consume the packages later inside Jetpack. But yeah, I think you're right, we eventually might have to end up adding the version specifically to `composer.json` of the packages if we want to consume the package as a symlink, and have a `composer.lock` inside Jetpack that's accurate about the package versions.

Reply:
To elaborate little more; I'd expect Jetpack depending on its own `/packages/**` be fluid but any transitive 3rd party dependencies in `/packages/*/composer.json` files to get locked down in some lockfile that's version controlled, to avoid surprises.

(So exactly like in Calypso)

Reply:
I see why that could be useful, but we still don't use any external libraries in our packages, and we might end up not doing that for a while. If we start doing it, and find the necessity, we can always start committing it. 

If folks feel strongly about this, we can commit it from the beginning - Composer's docs don't strictly forbid it.

Reply:
To be fair though, without actually depending on any of the packages in an external from Jetpack project, we will never know if it's useful or not to have `composer.lock` committed for packages. I say let's stay open and sharp-eyed for it, and see what comes ahead.